### PR TITLE
[wip] Barlow twins

### PR DESCRIPTION
**Pull request by ananyahjha93**
# Before submitting

Tutorial for SSL technique Barlow twins.

## What does this PR do?

Fixes # (issue).

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: lightning_examples/barlow-twins/barlow_twins.py [hunk 1]
```diff
@@ -0,0 +1,569 @@
...
+
+    def forward(self, z1, z2):
+        # N x D, where N is the batch size and D is output dim of projection head
+        z1_norm = (z1 - torch.mean(z1, dim=0)) / torch.std(z1, dim=0)
+        z2_norm = (z2 - torch.mean(z2, dim=0)) / torch.std(z2, dim=0)
+
+        cross_corr = torch.matmul(z1_norm.T, z2_norm) / self.batch_size
+
+        on_diag = torch.diagonal(cross_corr).add_(-1).pow_(2).sum()
+        off_diag = self.off_diagonal_ele(cross_corr).pow_(2).sum()
```
Comment (tchaton) [inline lightning_examples/barlow-twins/barlow_twins.py:153]: Might be cleaner, faster to multiple this with a mask instead. (cross_corr * (1 - torch.eye(self.batch_size)).pow_(2).sum()

File: lightning_examples/barlow-twins/barlow_twins.py [hunk 2]
```diff
@@ -0,0 +1,569 @@
...
+
+    def validation_step(self, batch, batch_idx):
+        loss = self.shared_step(batch)
+
+        self.log("val_loss", loss, on_step=False, on_epoch=True)
+        return loss
+
+    def configure_optimizers(self):
+        optimizer = torch.optim.Adam(self.parameters(), lr=self.learning_rate)
+
```
Comment (tchaton) [inline lightning_examples/barlow-twins/barlow_twins.py:536]: Not space needed. same above.

File: lightning_examples/barlow-twins/barlow_twins.py [hunk 3]
```diff
@@ -0,0 +1,569 @@
...
+    encoder_out_dim=encoder_out_dim,
+    num_training_samples=len(train_dataset),
+    batch_size=batch_size,
+    z_dim=z_dim,
+)
+
+trainer = Trainer(
+    max_epochs=max_epochs, gpus=torch.cuda.device_count(), precision=16 if torch.cuda.device_count() > 0 else 32
+)
+trainer.fit(model, train_loader, val_loader)
```
Comment (tchaton) [inline lightning_examples/barlow-twins/barlow_twins.py:569]: It is missing a section on evaluation and how to do downstream usage of the trained model.

**Comment by Borda** [comment]
@ananyahjha93 pls mars finished comments as resolved so we know what has left... :rabbit:

File: lightning_examples/barlow-twins/barlow_twins.py [hunk 4]
```diff
@@ -0,0 +1,569 @@
...
+
+# %% [markdown]
+# ### Barlow Twins Loss
+#
+# Here we define the loss function for Barlow Twins. It first normalizes the D dimensinonal vectors from the projection head and then computes the DxD cross-correlation matrix between the normalized vectors of the 2 views of each image.
+#
+# Then it splits this cross-correlation matrix into two parts. The first part, the diagonal of this matrix is brought closer to 1, which pushes up the cosine similarity between the latent vectors of two views of each image, thus making the backbone invariant to the transformations applied to the views. The second part of the loss pushes the non-diagonal elements of the cross-corrlelation matrix closes to 0. This reduces the redundancy between the different dimensions of the latent vector.
+
+# %%
+class BarlowTwinsLoss(nn.Module):
```
[the remark this answers is not in the document]
Comment (ananyahjha93) [reply lightning_examples/barlow-twins/barlow_twins.py:130]: I would prefer not to delve into loss landscape plots for this tutorial, as that would be a whole different topic to explain.

File: lightning_examples/barlow-twins/barlow_twins.py (hunk 1 above)
Comment (ananyahjha93) [reply lightning_examples/barlow-twins/barlow_twins.py:153]: Keeping this consistent with the official implementation: https://github.com/facebookresearch/barlowtwins/blob/main/main.py#L218

**Review by SeanNaren — APPROVED** [review]
Great notebook! maybe add a paper reference in the header somewhere?
